pages/map/index.tsx

```diff
@@ -48,52 +51,106 @@ const Marker = ({color}: MarkerProps) => (
         size={36}
         className={styles.marker}
     />
-)
+);
 
-const GMap = ({center, children}: GMapProps) => {
+const Map = ({center, children}: MapProps) => {
     const defaultCoords: Coords = {lat: 40.7128, lng: -74.0060};
     const mapOptions = {
         mapTypeControl: true,
         streetViewControl: true
     };
 
+    let isNull = Object.values(center).every(obj => obj === null);
+    
     return(
-        <>
-            {center.lat && center.lng ?
-                <div className={`${styles.map} shadow`}>
-                    <GoogleMapReact
-                        //@ts-ignore
-                        bootstrapURLKeys={{key: mapsKey}}
-                        defaultCenter={defaultCoords}
-                        center={center}
-                        defaultZoom={11}
-                        options={mapOptions}
-                    >
-                        <Marker
-                            //@ts-ignore
-                            lat={center.lat}
-                            lng={center.lng}
-                            color="blue"
-                        />
-                    </GoogleMapReact>
-                </div>
-                :
-                <LandingMessage/>
-            }
-        </>
+        <div className={`${styles.map} shadow`}>
+            <GoogleMap
+                //@ts-ignore
+                bootstrapURLKeys={{key: mapsKey, libraries: ["geometry", "drawing"]}}
+                onGoogleApiLoaded={({map, maps}) => {console.log(map, maps)}}
+                defaultCenter={defaultCoords}
+                center={isNull ? defaultCoords : center}
+                defaultZoom={11}
+                options={mapOptions}
+                yesIWantToUseGoogleMapApiInternals
+            >
+                {children}   
+            </GoogleMap>
+        </div>        
     );
 }
 
 const MapPageContent = () => {
     const [zip, setZip] = useState("");
     const [userCoords, setUserCoords] = useState({lat: null, lng: null});
+    const [places, setPlaces] = useState([]);
+    const [zipStatus, setZipStatus] = useState(false)
 
     const changeHandler = (e: any): void => {
         e.preventDefault();
-        console.log(zip);
         setZip(e.target.value);
     };
 
+    const submitHandler = (e: any) => {
+        e.preventDefault();
+        getPlaceCoordsWithZip();
+        setZip("");
+    }
+
+    const getPlaceCoordsWithUserCoords = async () => {
```

Comment:
one thing you can do to improve readability is to refactor some of these functions into [pure functions](https://en.wikipedia.org/wiki/Pure_function). for example, the async getters can be moved outside the component altogether which would condense the component code and make it easier to read. otherwise, it's hard to tell which variables are used by which functions and what state is modified as a result of executing those functions. it also makes the functions themselves harder to reuse because they're mixed with side effects.

- `getPlaceCoordsWithUserCoords` could be refactored into a helper method that takes `userCoords` and returns a list of places.
- `getPlaceCoordsWithZip` could be refactored into a helper that takes `zip` and returns the geocoded result.

then, the `useEffect` block below could be rewritten as:

```
useEffect(async () => {
  setPlaces(await getPlaceCoordsWithUserCoords(userCoords));
}, [userCoords]);
```

and the `submitHandler` above could be rewritten as:

```
const submitHandler = async (e: any) => {
  e.preventDefault();
  
  const geocodedResult = await getPlaceCoordsWithZip(zip);
  setZipStatus(true);

  setUserCoords(/* same code as before */);
}
```

with this, now you don't need to duplicate the place fetching logic since `setUserCoords` will cause the `useEffect` block to rerun.

Reply:
I gave the refactoring thing a try, where I called functions within the state setters. Doing that gave me runtime errors, where Next said something like ```(Object...) is undefined```
async callbacks & promises aren't a thing in useEffect because of race conditions apparently, but async functions can be called from within useEffect, which is pretty much what I'm already doing. Here's what I mean:

Bad:
`
useEffect(async () => {
})`

Good:

`const x = async () => {
        let z = await y
}`
`useEffect(() => {
        x()
})`

The same applies to using async event handlers like you suggested, although event.persist could probably be used. Here's some stuff I found about this:
[async useEffect](https://stackoverflow.com/questions/53332321/react-hook-warnings-for-async-function-in-useeffect-useeffect-function-must-ret)
[async useEffect GitHub issue](https://github.com/facebook/react/issues/14326)
[async event handlers](https://medium.com/@ian.mundy/async-event-handlers-in-react-a1590ed24399)

I could probably store the return values of the async getters in variables & then use the variables or properties of those variables to set states:

`const getPlacesWithUserCoords = async () => {foo: "bar", a: "b", c: [1, 2, 3, 4, 5]}`
`const placesReq = getPlacesWithUserCoords()`
`setPlaces(placesReq.c)`

Reply:
ah, that's too bad. I hate it when framework authors skimp on async design. what they should have done is created an API that makes it possible to cancel the previous effect if the effect is rerun before the async task completes. the reality is, you still have to deal with this problem regardless of whether the async task is passed directly to `useEffect` or run inside a sync callback.

regarding the runtime errors from Next, you should debug it some more to figure out why the error is occurring. I recommend [chrome dev tools](https://developers.google.com/web/tools/chrome-devtools/javascript) for debugging javascript. I also found an article specifically on [debugging Next](https://nextjs.org/docs/advanced-features/debugging). if you can't figure it out after an hour or so, we can go over it on Saturday. debugging is definitely a skill you'll want to get good at. I've spent just as many hours debugging as I have writing code. no one enjoys it, but it's a brute-force way of figuring out a mistake.

Reply:
Hey @stephenbunch . I was able to refactor the component. I moved all the API logic to a separate file altogether.  Everything is still working as intended, but there is now a delay in the marker loading, due to the async tasks. What I can do to combat this, as to not have a user leave the site, is display a loading spinner, and control it with a loading state. Either way, check this [commit](https://github.com/deybyr647/RecycleIT/pull/3/commits/fba720e253cb19f84dd6396a820eceee198fb4e0) whenever you can, and let me know what you think!